Add extra debug data to `cache_invalidate` log

convenient opened the pull request:
### Description (*)
Attach extra debug data to the `cache_invalidate:` log.  I recently came across another occurrence of this issue, but within a custom indexer https://github.com/magento/inventory/commit/b6d7223b8ca3b13c4049b3b0f6efc645b0848e01. I believe having extra information on hand would make peoples lives easier.

The indexer was erroneously purging all of redis several times, causing massive performance degradation. We can see frequent drops in redis performance on the web frontend.
<img width="500"  alt="redis-destroyed2" src="https://github.com/user-attachments/assets/bdf97baa-4bfa-4077-9134-e2a752416acf" />

And these are nicely correlated with the groupings of `cache_invalidate` with `{"tags":[],"mode":"all"}` purges.
<img width="500"  alt="redis-destroyed" src="https://github.com/user-attachments/assets/cb17ef5e-66b5-4e45-becb-9992029a2b56" />

In this case I knew what I was looking for because I was aware of https://github.com/magento/inventory/commit/b6d7223b8ca3b13c4049b3b0f6efc645b0848e01, but in cases like this I believe it would be neat if the Magento application just provided you with all the information to debug these scenarios up front. This change should make the system behave a bit nicer for both the core application in future (if any similar issues may occur) and for any custom development out there.

With this little bit of extra information, the stack trace will pinpoint the problem for the developers to take action without having to deploy similar logging or make guesswork. Given the impact, I think this is a sensible change. As far as I can see this flow is not really triggered normally in the application in any case.

### Manual testing scenarios (*)

**Cache flush**

`cache_invalidate` is not called at all, because a flush purges the whole storage separately.
```log
$ rm -f var/log/*; php bin/magento  cache:flush > /dev/null; grep cache_invalidate var/log/debug.log
grep: var/log/debug.log: No such file or directory
```

**Cache clean**

As per https://github.com/magento/magento2/issues/33122 `cache:clean` does not actually clean the whole Magento cache, it only purges the defined tags. So this log is not triggered
```log
$ rm -f var/log/*; php bin/magento  cache:clean > /dev/null; grep cache_invalidate var/log/debug.log
[2025-09-18T14:13:38.894538+00:00] main.DEBUG: cache_invalidate:  {"method":"GET","url":"http:/","invalidateInfo":{"tags":["CONFIG"],"mode":"matchingTag"}} []
[2025-09-18T14:13:38.897990+00:00] main.DEBUG: cache_invalidate:  {"method":"GET","url":"http:/","invalidateInfo":{"tags":["LAYOUT_GENERAL_CACHE_TAG"],"mode":"matchingTag"}} []
[2025-09-18T14:13:38.898470+00:00] main.DEBUG: cache_invalidate:  {"method":"GET","url":"http:/","invalidateInfo":{"tags":["COLLECTION_DATA"],"mode":"matchingTag"}} []
[2025-09-18T14:13:38.898723+00:00] main.DEBUG: cache_invalidate:  {"method":"GET","url":"http:/","invalidateInfo":{"tags":["REFLECTION"],"mode":"matchingTag"}} []
[2025-09-18T14:13:38.898971+00:00] main.DEBUG: cache_invalidate:  {"method":"GET","url":"http:/","invalidateInfo":{"tags":["DB_DDL"],"mode":"matchingTag"}} []
[2025-09-18T14:13:38.899276+00:00] main.DEBUG: cache_invalidate:  {"method":"GET","url":"http:/","invalidateInfo":{"tags":["COMPILED_CONFIG"],"mode":"matchingTag"}} []
[2025-09-18T14:13:38.899522+00:00] main.DEBUG: cache_invalidate:  {"method":"GET","url":"http:/","invalidateInfo":{"tags":["EAV"],"mode":"matchingTag"}} []
[2025-09-18T14:13:38.899769+00:00] main.DEBUG: cache_invalidate:  {"method":"GET","url":"http:/","invalidateInfo":{"tags":["CUSTOMER_NOTIFICATION"],"mode":"matchingTag"}} []
[2025-09-18T14:13:38.900013+00:00] main.DEBUG: cache_invalidate:  {"method":"GET","url":"http:/","invalidateInfo":{"tags":["INTEGRATION"],"mode":"matchingTag"}} []
[2025-09-18T14:13:38.900246+00:00] main.DEBUG: cache_invalidate:  {"method":"GET","url":"http:/","invalidateInfo":{"tags":["INTEGRATION_API_CONFIG"],"mode":"matchingTag"}} []
[2025-09-18T14:13:38.900497+00:00] main.DEBUG: cache_invalidate:  {"method":"GET","url":"http:/","invalidateInfo":{"tags":["GRAPHQL_QUERY_RESOLVER_RESULT"],"mode":"matchingTag"}} []
[2025-09-18T14:13:38.901090+00:00] main.DEBUG: cache_invalidate:  {"method":"GET","url":"http:/","invalidateInfo":{"tags":["WEBSERVICE"],"mode":"matchingTag"}} []
[2025-09-18T14:13:38.901351+00:00] main.DEBUG: cache_invalidate:  {"method":"GET","url":"http:/","invalidateInfo":{"tags":["TRANSLATE"],"mode":"matchingTag"}} []
```

**Additional test cases**
```php
//test.php
<?php
use Magento\Framework\App\Bootstrap;
require __DIR__ . '/app/bootstrap.php';

$bootstrap = Bootstrap::create(BP, $_SERVER);
$obj = $bootstrap->getObjectManager();
$logger = $obj->get(\Psr\Log\LoggerInterface::class);
$cache = $obj->get(\Magento\Framework\App\CacheInterface::class);

$logger->debug('cache_invalidate_test - cleaning one');
$cache->clean(['foobar']);

$logger->debug('cache_invalidate_test - cleaning multiple');
$cache->clean(['baz', 'qux']);

$logger->debug('cache_invalidate_test - cleaning no params');
$cache->clean();

$logger->debug('cache_invalidate_test - cleaning empty arrays');
$cache->clean([]);
```

Here's where we see the logs being triggered, in the scenario where `clean` is being called without parameters.

```log
$ rm -f var/log/*; php test.php > /dev/null; grep cache_invalidate var/log/debug.log
[2025-09-18T14:15:37.677293+00:00] main.DEBUG: cache_invalidate_test - cleaning one [] []
[2025-09-18T14:15:38.583136+00:00] main.DEBUG: cache_invalidate:  {"method":"GET","url":"http:/","invalidateInfo":{"tags":["foobar"],"mode":"matchingAnyTag"}} []
[2025-09-18T14:15:38.583175+00:00] main.DEBUG: cache_invalidate_test - cleaning multiple [] []
[2025-09-18T14:15:38.583294+00:00] main.DEBUG: cache_invalidate:  {"method":"GET","url":"http:/","invalidateInfo":{"tags":["baz","qux"],"mode":"matchingAnyTag"}} []
[2025-09-18T14:15:38.583312+00:00] main.DEBUG: cache_invalidate_test - cleaning no params [] []
[2025-09-18T14:15:38.583382+00:00] main.DEBUG: cache_invalidate:  {"method":"GET","url":"http:/","invalidateInfo":{"tags":[],"mode":"all"},"trace":[{"file":"/var/www/html/vendor/magento/framework/Cache/Frontend/Decorator/Logger.php","line":60,"function":"execute","class":"Magento\\Framework\\Cache\\InvalidateLogger","type":"->","args":[{"tags":[],"mode":"all"}]},{"file":"/var/www/html/vendor/magento/framework/Cache/Frontend/Decorator/Logger.php","line":48,"function":"log","class":"Magento\\Framework\\Cache\\Frontend\\Decorator\\Logger","type":"->","args":[{"tags":[],"mode":"all"}]},{"file":"/var/www/html/vendor/magento/framework/App/Cache.php","line":102,"function":"clean","class":"Magento\\Framework\\Cache\\Frontend\\Decorator\\Logger","type":"->","args":[]},{"file":"/var/www/html/vendor/magento/framework/App/Cache/Proxy.php","line":103,"function":"clean","class":"Magento\\Framework\\App\\Cache","type":"->","args":[[]]},{"file":"/var/www/html/test.php","line":19,"function":"clean","class":"Magento\\Framework\\App\\Cache\\Proxy","type":"->","args":[]}]} []
[2025-09-18T14:15:38.583606+00:00] main.DEBUG: cache_invalidate:  {"method":"GET","url":"http:/","invalidateInfo":{"tags":[],"mode":"all"},"trace":[{"file":"/var/www/html/vendor/magento/framework/Cache/Frontend/Decorator/Logger.php","line":60,"function":"execute","class":"Magento\\Framework\\Cache\\InvalidateLogger","type":"->","args":[{"tags":[],"mode":"all"}]},{"file":"/var/www/html/vendor/magento/framework/Cache/Frontend/Decorator/Logger.php","line":48,"function":"log","class":"Magento\\Framework\\Cache\\Frontend\\Decorator\\Logger","type":"->","args":[{"tags":[],"mode":"all"}]},{"file":"/var/www/html/vendor/magento/framework/App/Cache.php","line":102,"function":"clean","class":"Magento\\Framework\\Cache\\Frontend\\Decorator\\Logger","type":"->","args":[]},{"file":"/var/www/html/vendor/magento/framework/App/Cache/Proxy.php","line":103,"function":"clean","class":"Magento\\Framework\\App\\Cache","type":"->","args":[[]]},{"file":"/var/www/html/test.php","line":19,"function":"clean","class":"Magento\\Framework\\App\\Cache\\Proxy","type":"->","args":[]}]} []
[2025-09-18T14:15:38.583636+00:00] main.DEBUG: cache_invalidate_test - cleaning empty arrays [] []
[2025-09-18T14:15:38.583696+00:00] main.DEBUG: cache_invalidate:  {"method":"GET","url":"http:/","invalidateInfo":{"tags":[],"mode":"all"},"trace":[{"file":"/var/www/html/vendor/magento/framework/Cache/Frontend/Decorator/Logger.php","line":60,"function":"execute","class":"Magento\\Framework\\Cache\\InvalidateLogger","type":"->","args":[{"tags":[],"mode":"all"}]},{"file":"/var/www/html/vendor/magento/framework/Cache/Frontend/Decorator/Logger.php","line":48,"function":"log","class":"Magento\\Framework\\Cache\\Frontend\\Decorator\\Logger","type":"->","args":[{"tags":[],"mode":"all"}]},{"file":"/var/www/html/vendor/magento/framework/App/Cache.php","line":102,"function":"clean","class":"Magento\\Framework\\Cache\\Frontend\\Decorator\\Logger","type":"->","args":[]},{"file":"/var/www/html/vendor/magento/framework/App/Cache/Proxy.php","line":103,"function":"clean","class":"Magento\\Framework\\App\\Cache","type":"->","args":[[]]},{"file":"/var/www/html/test.php","line":22,"function":"clean","class":"Magento\\Framework\\App\\Cache\\Proxy","type":"->","args":[[]]}]} []
[2025-09-18T14:15:38.583829+00:00] main.DEBUG: cache_invalidate:  {"method":"GET","url":"http:/","invalidateInfo":{"tags":[],"mode":"all"},"trace":[{"file":"/var/www/html/vendor/magento/framework/Cache/Frontend/Decorator/Logger.php","line":60,"function":"execute","class":"Magento\\Framework\\Cache\\InvalidateLogger","type":"->","args":[{"tags":[],"mode":"all"}]},{"file":"/var/www/html/vendor/magento/framework/Cache/Frontend/Decorator/Logger.php","line":48,"function":"log","class":"Magento\\Framework\\Cache\\Frontend\\Decorator\\Logger","type":"->","args":[{"tags":[],"mode":"all"}]},{"file":"/var/www/html/vendor/magento/framework/App/Cache.php","line":102,"function":"clean","class":"Magento\\Framework\\Cache\\Frontend\\Decorator\\Logger","type":"->","args":[]},{"file":"/var/www/html/vendor/magento/framework/App/Cache/Proxy.php","line":103,"function":"clean","class":"Magento\\Framework\\App\\Cache","type":"->","args":[[]]},{"file":"/var/www/html/test.php","line":22,"function":"clean","class":"Magento\\Framework\\App\\Cache\\Proxy","type":"->","args":[[]]}]} []
```

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)


### Resolved issues:
1. [x] resolves magento/magento2#40204: Add extra debug data to `cache_invalidate` log

convenient committed:
Add extra debug data to `cache_invalidate` log

kandy commented on lib/internal/Magento/Framework/Cache/InvalidateLogger.php:
```diff
@@ -43,7 +43,15 @@ public function __construct(HttpRequest $request, Logger $logger)
+                $context['trace'] =  (new \Exception)->getTrace();
+            }
+        }
+        $this->logger->debug('cache_invalidate: ', $context);
```
btw, if. you pass exception to psr log based on monolog in format like $this->logger->debug('cache_invalidate: ', ['exception'=> new \Exception()]); it will add trace to message

convenient replied:
Ooh i didn't know that thanks. I've just given it a quick test like this to see
```php
$this->logger->debug('cache_invalidate: ', ['exception'=> new \Exception()]);
```

and i get this log, it does the exception but without the full trace
```log
main.DEBUG: cache_invalidate:  {"exception":"[object] (Exception(code: 0):  at /var/www/html/vendor/magento/framework/Cache/InvalidateLogger.php:54)"}
```

Am I missing a trick?

rhoerr replied:
Can't speak to @kandy's message, but another option is casting the exception to string. That should give the message, code, and full formatted trace. `(string)(new \Exception())` (see https://www.php.net/manual/en/exception.tostring.php)

convenient replied:
@rhoerr that's decent, looks like this

```php
public function execute($invalidateInfo)
{
    $context = $this->makeParams($invalidateInfo);
    if (isset($invalidateInfo['tags'], $invalidateInfo['mode'])) {
        if ($invalidateInfo['mode'] === 'all' && is_array($invalidateInfo['tags']) && empty($invalidateInfo['tags'])) {
            // If we are sending a purge request to all cache storage capture the trace
            // This is not a usual flow, and likely a bug is causing a performance issue
            $context['exception'] =  (string)(new \Exception('full purge of cache storage triggered'));
        }
    }
    $this->logger->debug('cache_invalidate: ', $context);
}
```

```log
[2025-09-18T15:50:52.353321+00:00] main.DEBUG: cache_invalidate:  {"method":"GET","url":"http:/","invalidateInfo":{"tags":[],"mode":"all"},"exception":"Exception: full purge of cache storage triggered in /var/www/html/vendor/magento/framework/Cache/InvalidateLogger.php:51
Stack trace:
#0 /var/www/html/vendor/magento/framework/Cache/Frontend/Decorator/Logger.php(60): Magento\\Framework\\Cache\\InvalidateLogger->execute(Array)
#1 /var/www/html/vendor/magento/framework/Cache/Frontend/Decorator/Logger.php(48): Magento\\Framework\\Cache\\Frontend\\Decorator\\Logger->log(Array)
#2 /var/www/html/vendor/magento/framework/App/Cache.php(102): Magento\\Framework\\Cache\\Frontend\\Decorator\\Logger->clean()
#3 /var/www/html/vendor/magento/framework/App/Cache/Proxy.php(103): Magento\\Framework\\App\\Cache->clean(Array)
#4 /var/www/html/test.php(22): Magento\\Framework\\App\\Cache\\Proxy->clean(Array)
#5 {main}"} []
```

I'm happy with any approach really, so long as we get the full stack trace to work with. Can the adobe team please advise what they prefer :)

convenient committed:
Fix line length

convenient committed:
Update copyright header

https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/40196/0b62a7b8683ef732e82264c00c926c60/Statics/report-sanity-ee.html
```
/**
 * Copyright {{CREATION_YEAR}} Adobe
 * All Rights Reserved.
 */

```

convenient committed:
Update InvalidateLogger.php

See https://github.com/magento/magento2/pull/40196#discussion_r2360022200

convenient committed:
Merge branch '2.4-develop' into patch-3

convenient committed:
Update InvalidateLogger.php

engcom-Hotel committed:
Merge branch '2.4-develop' into patch-3

engcom-Hotel commented:
@magento run all tests

engcom-Hotel reviewed (changes requested):
Thank you @convenient for this PR! This is a valuable improvement for debugging cache performance issues.

Regarding @kandy's comment about exception handling [here](https://github.com/magento/magento2/pull/40196#discussion_r2359659185) - they're absolutely right, and this would make the implementation even better.

**Suggested change:**

Instead of:
```
$context['exception'] = (string)(new \Exception('full purge of cache storage triggered'));
```

Use:
```
$context['exception'] = new \Exception('full purge of cache storage triggered');
```

This leverages Monolog's native exception handling per PSR-3 standard.

convenient commented:
Hey @engcom-Hotel can you please see my comment https://github.com/magento/magento2/pull/40196#discussion_r2359678518

When I used that mechanism it did not contain the full trace, and therefore wouldnt be able to be used to help fix such issues

engcom-Hotel reviewed (approved):
Thank you @convenient for your reply!

I am approving this PR for further processing.

engcom-Dash committed:
Merge branch '2.4-develop' into patch-3

engcom-Dash committed:
Merge branch '2.4-develop' into patch-3
